graphql_compiler/compiler/emit_sql.py

```diff
@@ -205,23 +259,60 @@ def _set_group_by(self, group_by):
         """Set output columns for the fold object."""
         self._group_by = group_by
 
-    def _construct_fold_joins(self, edge, from_alias, to_alias):
+    def _construct_fold_joins(self):
         """Use the edge descriptors to create the join clause between the tables in the fold."""
-        join_clause = sqlalchemy.join(
-            from_alias,
-            to_alias,
-            onclause=(from_alias.c[edge.from_column] == to_alias.c[edge.to_column])
-        )
+        # Start the join clause with the from_table of the first join descriptor
+        _, join_clause, _ = self.join_info[-1]
+
+        # Starting at the first from_table, join traversed vertices until the output vertex is
+        # reached.
+        #
+        # MSSQL doesn't use the last JOIN so we pop it off before construction the JOIN clause
+        if self._dialect == 'mssql':
+            # Move the final join predicate (from the vertex preceding the output to
+            # the output vertex) into the WHERE clause of the SELECT ... FOR XML PATH('')
+            self.join_info.pop()
+        while len(self.join_info) > 0:
+            # more recent joins are at the end of the list
+            edge, from_alias, to_alias = self.join_info.pop()
+            join_clause = sqlalchemy.join(
+                join_clause,
+                to_alias,
+                onclause=(from_alias.c[edge.from_column] == to_alias.c[edge.to_column])
+            )
         return join_clause
 
     def _construct_fold_subquery(self, subquery_from_clause):
         """Combine all parts of the fold object to produce the complete fold subquery."""
-        return sqlalchemy.select(
+        select_statement = sqlalchemy.select(
             self.outputs
         ).select_from(
             subquery_from_clause
-        ).group_by(
-            *self.group_by
+        )
+
+        if self._dialect == 'mssql':
+            # mssql 2014 doesn't rely on a group by
+            return select_statement
+        else:
+            return select_statement.group_by(
+                *self.group_by
+            )
+
+    def _append_default_column(self, intermediate_fold_output_name, fold_output):
+        """Select an array_agg of the fold output field, labeled as requested."""
+        self._outputs.append(
```

Comment:
Rather than state-mutating functions like this, why not restructure this into a stateless function that does the `sqlalchemy.func.array_agg().label()` magic, and then take its output and append it to `self._outputs` in the caller function?